### PR TITLE
Particles: Do not send single particles to distant clients 

### PR DESCRIPTION
**Pull request by paramat**
Previously, every individual particle on a server is sent to, and rendered by
(even if not actually visible), every client regardless of distance. This
significantly reduces client FPS and creates unnecessary network traffic.
Maximum distance is set by 'max block send distance' as this determines how far
a client is able to see.
/////////////////////////////////

Tested.
Partial fix for #5667 work on particle spawners comes later.
I have used code https://github.com/minetest/minetest/blob/master/src/server.cpp#L2079 from 'play sound' to limit distance.
By default the max distance is 160 nodes (max block send distance = 10).

### CODE REVIEW

File: src/server.cpp [hunk 1]
```diff
@@ -1708,12 +1708,21 @@ void Server::SendSpawnParticle(u16 peer_id, u16 protocol_version,
...
 			RemotePlayer *player = m_env->getPlayer(*i);
 			if (!player)
 				continue;
+
+			PlayerSAO *sao = player->getPlayerSAO();
+			if (!sao)
+				continue;
+
+			if (sao->getBasePosition().getDistanceFrom(pos) >
+					160.0f)
```
Comment (nerzhul) [inline src/server.cpp]: can you set a define for this ?

**Comment by Fixer-007** [comment]
I see this hardcoded 160.0f, maybe better utilise "max_block_send_distance" in calculation?

File: src/server.cpp (hunk 1 above)
Comment (numberZero) [inline src/server.cpp]: @sfan5 suggested to use `active_block_range` instead https://github.com/minetest/minetest/issues/5667#issuecomment-297980867

**Comment by paramat** [comment]
Fixer-007

> which is currently hardcoded to 160.0 nodes for testing.

> Add the check to Server::SendAddParticleSpawner. AFAIK that’s what need to be limited, Server::SendSpawnParticle doesn’t seem to be used widely

Yes it's WIP, trying to get it working for 'add particle' first before duplicating code. Single particles are used a lot.

**Comment by paramat** [comment]
Updated and tested, seems to work now, please test.
Still 'addparticle' only (test with my snowdrift mod) and hardcoded to 160 nodes for testing, i can add a new setting or use another setting for this distance later.

**Comment by Fixer-007** [comment]
Will test

**Comment by paramat** [comment]
Added to milestone as Fixer's testing shows a large performance reduction due to current behaviour, note the comments in the current code saying 'this sucks'.

**Comment by raymoo** [comment]
Not sending particle spawners is a problem since they tend to be longer lived than particles. If I walk into the range of a particle spawner I should be able to see it.

Send the particle spawner always, but just don't generate particles if it's too far away. This will still solve the performance problems because they are caused by particle rendering, not particle spawners themselves.

You could also send and spawn all particles and just not render far away particles. This is more expensive than switching off spawners, but should still solve the issue since the distance check would be much much cheaper than rendering a particle. This way would preserve the most particle functionality.

**Comment by nerzhul** [comment]
@paramat i agree with @raymoo please remove the limit on particlespawner, we can fix that after release adding a better algorithm to detect if a particle spawner should be sent or not

**Comment by Fixer-007** [comment]
Oh... it is server side... I will check tnt explosion and rain mod.

**Comment by paramat** [comment]
> Not sending particle spawners is a problem since they tend to be longer lived than particles. If I walk into the range of a particle spawner I should be able to see it.

Send the particle spawner always, but just don't generate particles if it's too far away. This will still solve the performance problems because they are caused by particle rendering, not particle spawners themselves.

Good idea. The limit was planned for arouond 160 nodes but some particle spawners could be very long lived. I'll rework particle spawners.

**Comment by paramat** [comment]
Updated to only affect single particles, spawners can be another PR, as it might be more complex and release is close.
This nnow uses 'max block send distance'.
@raymoo

**Comment by paramat** [comment]
Retested on a locel server and client setup and confirmed particle cutoff is exactly as set by max_block_send_distance.

**Comment by nerzhul** [comment]
Okay for me. Removing particle spawner limit is reasonable, this was not in our roadmap nor a bug, it was there since the beginning we can delay the fix for that from this release.

**Comment by paramat** [comment]
Thanks, the PR for particle spawners is here #5766

**Pull request merged**